File: src/main/kotlin/com/fantastictrio/cw4sem/model/Organization.java
package com.fantastictrio.cw4sem.model;

import lombok.Data;

import javax.persistence.*;
import java.util.List;

@Data
@Entity(name = "organization")
public class Organization {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    @Column(nullable = false, unique = true)
    private String name;

    @Column(nullable = false)
    private String type;

    @OneToMany(mappedBy = "organization",cascade=CascadeType.PERSIST)
    private List<User> users;

    @Override
    public String toString() {
        return "Organization{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", type='" + type + '\'' +
                ", users=" + users +
                '}';
    }
}

File: src/main/kotlin/com/fantastictrio/cw4sem/model/Organization.kt
package com.fantastictrio.cw4sem.model

import com.fantastictrio.cw4sem.dto.OrganizationPayload
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import org.hibernate.annotations.Nationalized
import java.time.Instant
import javax.persistence.*
import javax.validation.constraints.Pattern

@JsonIgnoreProperties(value = ["users"])
@Entity(name = "organization")
data class Organization(

    @Nationalized
    @Pattern(regexp = "[a-zA-Zа-яА-Я\\s\\d]{3,30}")
    @Column(nullable = false, unique = true)
    val name: String,

    @Nationalized
    @Pattern(regexp = "[a-zA-Zа-яА-Я\\s]{3,30}")
    @Column(nullable = false)
    val type: String,

    @OneToMany(mappedBy = "organization", cascade = [CascadeType.PERSIST])
    val users: List<User> = emptyList(),

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Int = 0,

    @Column(nullable = false)
    val createdDate: Instant = Instant.now(),
) {
    constructor(payload: OrganizationPayload, id: Int = 0, users: List<User> = emptyList()) : this(
        payload.name,
        payload.type,
        users,
        id,
    )

    fun toPayload(): OrganizationPayload {
        return OrganizationPayload(this)
    }

    val employeeCount: Int
    get() = users.size
}
